### PR TITLE
Bring gnat6 back

### PR DESCRIPTION
**Pull request by mmlb**
###### Description of changes

Re-enables gnat6. This is mostly so we can build coreboot-4.11 without having to override its own build-tools scripts. coreboot seems to be very picky about its compilers and I'd like to let it do its thing. This will be helpful for https://github.com/osresearch/heads/pull/1269 for example.

I've also renamed gnatboot to gnat-bootstrap. I did this primarily for 2 reasons.

1. Most of the other boostrap packages are named $compiler-bootstrap so it makes sense to keep that pattern.
2. As someone not super knowledgeable about GCC's Ada and co I originally assumed `gnatboot` was the name of an actual binary, not a bootstrapping compiler. The new names makes it much more obvious to others like me.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mmlb** [comment]
@vcunat for you to :eye: too. Thanks for your pointer in matrix!

**Comment by tlaurion** [comment]
Adding some background notes on why this is badly needed.

Gnat requires to be built from older versions. Gnat is required to build coreboot's libgfxinit, that is, native graphic initialization, providing firmware based text console or framebuffer.

Coreboot buildstack is cross-built, per coreboot version, from host's buildstack to provide per version reproducible bootblock, ramstage, romstage etc. But for one to be able to build coreboot'gnat, host buildstack needs to provide older version of gnat. The host buildstack is also used to build musl-cross-make under Heads.

So without gnat6, coreboot 4.11 platforms for which a configuration requires gnat will fail. Newer coreboot buildstack would succeed being built, but not older. And older buildstack are only needed to build newer buildstack.

NixOS already builds coreboot buildstack for newer coreboot versions (4.18+ if I recall well) but older and freer platforms having been dropped by coreboot (kgpe-d16, kcma-d8) under 4.11 won't be possible to build on NixOS without gnat6.

Can't wait to see this merged and starting to collaborate on building reproducible docker images and recipes to build on top of other Linux OS, dodging other reproducibility issues linked to having so many Linux host tools being moving targets and have the same ROMs built for same commit 10 years from now.

If one wants to buol

File: pkgs/development/compilers/gnat-bootstrap/default.nix [hunk 1]
```diff
@@ -38,15 +47,24 @@ let
...
+  src =
+    if majorVersion == "4" then
+      fetchurl {
+        inherit url hash;
+      }
+    else
+      fetchzip {
+        url = "https://github.com/alire-project/GNAT-FSF-builds/releases/download/gnat-${version}/gnat-${stdenv.hostPlatform.system}-${version}.tar.gz";
+        inherit hash;
+      };
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gnat-bootstrap/default.nix:67]: Maybe we can move this into versionMap, to reduce the branching and make the code a bit cleaner.

**Comment by RaitoBezarius** [comment]
I requested reviews from @felixsinger and @blitz as you know you have interest or have been involved with coreboot ecosystem in the past. Apologies if this was not desired.

**Comment by RaitoBezarius** [comment]
This is an awesome PR as I use coreboot on my X230 with NixOS and I have been looking into adding a coreboot NixOS module to manage end-to-end everything in NixOS. :)

Some classical points: please read the CONTRIBUTING guide, rename your commits accordingly `gnat6: init`, etc.

**Review by RaitoBezarius — APPROVED** [review]
This looks good to me; I didn't test it yet though.

**Comment by yu-re-ka** [comment]
> This is mostly so we can build coreboot-4.11 without having to override its own build-tools scripts.

May I ask why you are not building a more recent of coreboot?

**Comment by felixsinger** [comment]
@RaitoBezarius Thanks for mentioning :) Will try to have a look later in the evening.

**Comment by RaitoBezarius** [comment]
> > This is mostly so we can build coreboot-4.11 without having to override its own build-tools scripts.
> 
> May I ask why you are not building a more recent of coreboot?

This is mentioned in the second message:

> NixOS already builds coreboot buildstack for newer coreboot versions (4.16+ if I recall well) but older and freer platforms having been dropped by coreboot (kgpe-d16, kcma-d8) under 4.11 won't cross compile on NixOS without gnat6.

**Comment by alyssais** [comment]
> NixOS already builds coreboot buildstack for newer coreboot versions (4.16+ if I recall well) but older and freer platforms having been dropped by coreboot (kgpe-d16, kcma-d8) under 4.11 won't cross compile on NixOS without gnat6.

Are these older coreboot versions being maintained?  What would it take to update their required gnat version?  I'm sympathetic, but I'm not sure we want to commit to indefinitely maintaining every package that happens to be required by an old, fixed version of Coreboot.

File: pkgs/development/compilers/gnat-bootstrap/default.nix [hunk 2]
```diff
@@ -78,15 +96,22 @@ stdenv.mkDerivation rec {
...
 
   installPhase = ''
     mkdir -p $out
     cp -ar * $out/
+
+    if [[ ${majorVersion} == 4 ]]; then
+      mv $out/bin/gnatgcc_2wrap $out/bin/gnatgcc
+      ln -s $out/bin/gnatgcc $out/bin/gcc
+    fi
   ''
```
[the remark this answers is not in the document]
Comment (mmlb) [reply pkgs/development/compilers/gnat-bootstrap/default.nix:112]: can do

File: pkgs/development/compilers/gnat-bootstrap/default.nix (hunk 1 above)
Comment (mmlb) [reply pkgs/development/compilers/gnat-bootstrap/default.nix:67]: I tried this there too originally, the `version` is easily moved up and I'll add that in. On the other hand `src` is quite a pain. I'm not sure how I'd avoid repeating the alire-project releases url for each version and architecture, it doesn't seem to end up being cleaner at all for `src`.

**Comment by yu-re-ka** [comment]
+1, I would be much happier if we could figure out how to build coreboot for your platform without keeping around the older compiler(s)

File: pkgs/development/compilers/gnat-bootstrap/default.nix (hunk 1 above)
Comment (mmlb) [reply pkgs/development/compilers/gnat-bootstrap/default.nix]: Actually that doesn't work either, not sure why. When I tried building with cb798bcae40 nix-build errored with:
```
nix-build -A gnat-bootstrap4
error: undefined variable 'gccVersion'

       at /home/manny/cloned/github.com/NixOS/nixpkgs/pkgs/development/compilers/gnat-bootstrap/default.nix:23:20:

           22|       alireRevision = "4";
           23|       version = "${gccVersion}-${alireRevision}";
             |                    ^
           24|     } // {
```

So going to go back.

**Comment by mmlb** [comment]
I can't really say whether we *should* have this in nixpkgs or not. I suspect it would be very welcome by anyone wanting to build coreboot for their platform. I suspect the maintenance burden to be low. I'd also like to try and use gnat6 as the bootstrap compiler for the other gnats. It would be nice to minimize external dependencies imo.

**Comment by mmlb** [comment]
> > I suspect it would be very welcome by anyone wanting to build coreboot for their platform
> 
> You mean for platforms that have been removed in the latest coreboot version

Yep.

> Can you send me the nix expression you are using to build coreboot 4.11? I would be happy to help with making it build with a more recent gnat

I'm not using coreboot myself. I was first interested in this work to enable osresearch/heads to use nixpkgs for their build environment, which wants to keep making use of coreboot's compiler build scripts which need gnat6 in order to compile gnat6.

**Comment by alyssais** [comment]
> I'd also like to try and use gnat6 as the bootstrap compiler for the other gnats. It would be nice to minimize external dependencies imo.

Agreed that it would be good to bootstrap it.  What does the bootstrap chain for a current gnat look like?

**Comment by ethindp** [comment]
When the old GNAT 6 setup was used, it never was able to fully bootstrap for me. It always broke when it tried to bootstrap GCC 8/9. From what I understood from the mess, it went something like: gnatboot -> gnat6 -> gnat8 -> gnat9 -> gnat10 -> gnat11 -> gnat12 -> gnat13. My fix for this significantly reduced the convoluted mess of that process, though I now see that it broke things. Gnatboot was gnat 4, specifically from Gentoo. Is there a way we can do this without creating a combinatorial explosion?

**Comment by tlaurion** [comment]
@ethindp 

> My fix for this significantly reduced the convoluted mess of that process, though I now see that it broke things.

I'm sorry, getting hands on NixOS here myself. What happened, what broke and why gnat6 got dropped is a mystery to me still today from what I understand of NixOS.


What Heads would want, here, is to create a reproducible docker image to use current Make buildsystem which requires gnat6 (and older build toolchain) to bootstrap musl-cross-make's gcc (old but we don't really care until we do per newer tools requirement) and to bootstrap coreboot's version buildstack. From current experience, there is no problem doing that from older buildstack, the problem is the other way around. If one pops a debian-10 docker today, with its own latest make, awk (those two were problematic back then and required patches to build local make and local awk), then it would be possible to build coreboot 4.11 buildstack from gnat6 provided there. If make and awk (2+) were there, all other newer coreboot buildstacks could be built with older ones. So if NixOS provides gnat6, gnat6 should be able to build coreboot 4.11+ buildstack and current pinned musl-cross-make used to build every other modules.

As I understand it, gnat6 could also be used to build newer gnat under NixOS. But I might miss something here. I that your concern?

**Comment by RaitoBezarius** [comment]
> @ethindp
> 
> > My fix for this significantly reduced the convoluted mess of that process, though I now see that it broke things.
> 
> I'm sorry, getting hands on NixOS here myself. What happened, what broke and why gnat6 got dropped is a mystery to me still today from what I understand of NixOS.

FWIW, we have a policy of cleaning up old compilers (), gnat6 was dropped in 283b4c25f416ecf75d53afa108e9753e57ba3d66 / #210564 with gnat9, gnat10.

The rationale was the following:

> They haven't succeeded for half a year and people don't seem
> to miss them, so let's reduce the clutter on Hydra, etc.
> https://hydra.nixos.org/job/nixpkgs/trunk/gnat10.x86_64-linux/all

So it's a maintenance thing.

**Comment by tlaurion** [comment]
> I suspect it would be very welcome by anyone wanting to build coreboot for their platform
> 
> You mean for platforms that have been removed in the latest coreboot version
> 
> Can you send me the nix expression you are using to build coreboot 4.11? I would be happy to help with making it build with a more recent gnat

Ownerboot is building coreboot specific buildstack by patching older version build stack to conform with newer host buildstack, but that will fail later on when deprecating rules will hit again, if I understand well. Ownerboot then tells coreboot, through coreboot config option, to not use coreboot version's specific buildstack, but to use the one exposed on the host

  https://git.sr.ht/~amjoseph/ownerboot/tree/aa61ed1bd20ab415d678003f23834d31cd620457/item/src/coreboot-toolchain

**Comment by RaitoBezarius** [comment]
cc @amjoseph-nixpkgs as he seems involved in those shenigans :>.

**Comment by tlaurion** [comment]
@felixsinger any insights here?

**Comment by amjoseph-nixpkgs** [comment]
> cc @amjoseph-nixpkgs as he seems involved in those shenigans :>.

Actually coreboot only needs gnat if you want VGA output on x86 before handoff to the Linux kernel.  Just that one feature uses GNAT; nothing else does.  So I disabled that feature quite some time ago (serial consoles are better for remote access anyways) so that fighting with GNAT wouldn't keep eating up so much of my time.

But I support the goals of this PR!

**Comment by amjoseph-nixpkgs** [comment]
> When the old GNAT 6 setup was used, it never was able to fully bootstrap for me. It always broke when it tried to bootstrap GCC 8/9.

Yeah, I remember this too.  It was pretty bad.  Nixpkgs still had gnat6 when I first started using it to build coreboot, but I remember a lot of painful battles fought with it.  Enabling `langAda` (which is off by default) means building two interlocking bootstrap chains (gcc and gnat at each step).

**Comment by amjoseph-nixpkgs** [comment]
> Are these older coreboot versions being maintained?

NB: this is the open source version of "[have you stopped beating your dog](https://en.wikipedia.org/wiki/Loaded_question)?"

Briefly: non-UEFI bootloaders don't have an API; they aren't part of an ecosystem.  They do one simple job and get out of the way -- like your keyboard's firmware, which is probably unmaintained too (if it isn't `pkgs.qmk`).


> but I'm not sure we want to commit to indefinitely maintaining every package that happens to be required by an old, static version of Coreboot.

This is totally valid and reasonable objection.

But I really think that this is a GNAT-specific problem, not a coreboot problem.  However the problem isn't needing an old toolchain.  The problem is that *the old toolchain can't be built by a **new** toolchain*.


> we have a policy of cleaning up old compilers

*... if they can't be built by the latest gcc*.  We still ship `gcc-6` and `gcc-arm-embedded-6`.


> What would it take to update their required gnat version?

@tlaurion has anybody tried backporting the libgfxinit (the only gnat-using part of coreboot) from a recent coreboot back to 4.11?  Since it's written in a completely different language (Ada) than the rest of coreboot (C) the interfaces should be well-defined.

<details><summary>Side tangent about coreboot-4.11 maintenance branches</summary>
One of the headaches here is that although upstream has a branch for 4.11 they don't seem to be accepting patches to it (two commits in the last two years even though many people are working on this branch).  I have several small, useful, and easily-reviewable bugfixes that I would like to submit when I finally upgrade from 4.9 to 4.11 (which is the terminal version for non-IntelME/AmdPSP hardware).  I don't really know where to submit them.  Dasharo's fork is sort of a mass-rewrite so they don't apply (and switching to that is too much of a leap for me right now).
</details>

**Review by ghost — COMMENTED** [review]
Could you submit [gnatboot: rename to gnat-bootstrap](https://github.com/NixOS/nixpkgs/pull/225191/commits/d8dca4a430440c1a348658a82f67a2b84245cb52) as a separate PR?  It seems uncontroversial and causes no rebuilds.

File: pkgs/development/compilers/gnat-bootstrap/default.nix [hunk 3]
```diff
@@ -38,15 +47,25 @@ let
...
-  inherit gccVersion alireRevision;
+  pname = "gnat-bootstrap";
+  inherit gccVersion;
 
-  version = "${gccVersion}-${alireRevision}";
+    version =
+    if majorVersion == "4" then
+      gccVersion
+    else
+      "${gccVersion}-${alireRevision}";
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gnat-bootstrap/default.nix:57]:
```suggestion
  version = "${gccVersion}${alireRevision}";
```

**Comment by ethindp** [comment]
> > When the old GNAT 6 setup was used, it never was able to fully bootstrap for me. It always broke when it tried to bootstrap GCC 8/9.
> Yeah, I remember this too. It was pretty bad. Nixpkgs still had gnat6 when I first started using it to build coreboot, but I remember a lot of painful battles fought with it. Enabling langAda (which is off by default) means building two interlocking bootstrap chains (gcc and gnat at each step).

It was very bad. It made using gnat for anything but Coreboot impossible. Which is why I replaced that entire mess with what I believed to be, at the time, a simpler bootstrap chain: go straight from gnat 11 onwards. Sure, you still had to potentially bootstrap several GCCs, but that, at least, was one of those "we know it works" setups. Then it broke Coreboot (along with several other projects), and I had no idea precisely how to resolve the problem.

**Review by ghost — COMMENTED** [review]
Some of my review comments were really hard to express via github's UI, so I put them in a separate PR.  Could you please rebase this on top of it?

- https://github.com/NixOS/nixpkgs/pull/226179

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -14968,6 +14968,22 @@ with pkgs;
...
+    langC = true;
+    langCC = false;
+    langAda = true;
+    profiledCompiler = false;
+    # As per upstream instructions building a cross compiler
+    # should be done with a (native) compiler of the same version.
+    # If we are cross-compiling GNAT, we may as well do the same.
+    gnat-bootstrap =
+      if stdenv.hostPlatform == stdenv.targetPlatform
+         && stdenv.buildPlatform == stdenv.hostPlatform
```
Comment (amjoseph-nixpkgs) [inline pkgs/top-level/all-packages.nix:14982]:
```suggestion
         && stdenv.buildPlatform.canExecute stdenv.hostPlatform
```

File: pkgs/development/compilers/gnat-bootstrap/default.nix [hunk 5]
```diff
@@ -70,23 +89,29 @@ stdenv.mkDerivation rec {
...
+
+    # The included fixincl binary that is called during header fixup has a
+    # hardcoded execvp("/usr/bin/sed", ...) call, but /usr/bin/sed isn't
+    # available in the Nix Darwin stdenv.  Fortunately, execvp() will search the
+    # PATH environment variable for the executable if its first argument does not
+    # contain a slash, so we can just change the string to "sed" and zero the
+    # other bytes.
+    if [[ -f libexec/gcc/${upstreamTriplet}/${gccVersion}/install-tools/fixincl ]]; then
+      sed -i "s,/usr/bin/sed,sed\x00\x00\x00\x00\x00\x00\x00\x00\x00," libexec/gcc/${upstreamTriplet}/${gccVersion}/install-tools/fixincl
+    fi
```
[the remark this answers is not in the document]
Comment (mmlb) [reply pkgs/development/compilers/gnat-bootstrap/default.nix:101]: Thanks this makes a lot of sense, I wasn’t sure why it’s was separated before.

**Comment by amjoseph-nixpkgs** [comment]
I merged #226366, so if you rebase to `origin/master` the size of the diff will drop dramatically.

**Comment by tlaurion** [comment]
@mmlb did you have a time to look at this/rebase?

**Comment by mmlb** [comment]
@amjoseph-nixpkgs, I've rebased on top of #226179 to see what that looks like for this PR.

**Comment by tlaurion** [comment]
@amjoseph-nixpkgs ping :)

**Comment by tlaurion** [comment]
Ada might still be required. Otherwise some platforms stay in the dark until DRM+GPU kernel drivers take ownership of fb. With coreboot libgfxinit (which requires ada/gnat) a bootsplash can be showed within milliseconds until DRM+GPU takes over fb to provide proper fbwhiptail support.

Alternatives are evaluated, but gnat support should not be excluded, at least in the transition.

Details https://github.com/osresearch/heads/pull/1381#issuecomment-1542610357

**Comment by alyssais** [comment]
I feel like I'm still missing answers to some important questions here.

1. Would this be a useful step towards obtaining a binary-free bootstrap of modern GNAT?  If so, how far away are we from such a bootstrap?  That would sidestep my concerns about packaging compilers only used by unmaintained software, because it provides another reason to keep it around.
2. > @tlaurion has anybody tried backporting the libgfxinit (the only gnat-using part of coreboot) from a recent coreboot back to 4.11? Since it's written in a completely different language (Ada) than the rest of coreboot (C) the interfaces should be well-defined.

**Review by ghost — APPROVED** [review]
LGTM.  Sorry about the delayed response.

**Comment by mmlb** [comment]
> I feel like I'm still missing answers to some important questions here.
> 
>     1. Would this be a useful step towards obtaining a binary-free bootstrap of modern GNAT?  If so, how far away are we from such a bootstrap?  That would sidestep my concerns about packaging compilers only used by unmaintained software, because it provides another reason to keep it around.

No, this is still using the same gnat4 binary originally from gentoo. If binary free bootstrapping is wanted I'm happy to look into it for gnat.

**Comment by tlaurion** [comment]
> 2\. @tlaurion has anybody tried backporting the libgfxinit (the only gnat-using part of coreboot) from a recent coreboot back to 4.11? Since it's written in a completely different language (Ada) than the rest of coreboot (C) the interfaces should be well-defined.

@alyssais Sorry for the delay answering this.
The problem with upstream deprecated coreboot branches is not really the libgfxinit code itself but the buildstack pinned per coreboot versions to bootstrap builds for reproducibility. Those version numbered branches rarely gets patches upstreamed, which stall under review.coreboot.org. 

As said in tickets referred here, libgfxinit is not "required" anymore on the only platform having GMA (which libgfxinit supports) while kgpe-d16 implements native graphic init and doesn't need libgfxinit at all.

The real problem here, if distributions wants to use Nix/NixOS as a buildstack for coreboot branches, is that coreboot by itself, outside of its possible configurations, will not build unless gnat is available for libgfxinit. So for example here, libreboot being a distribution of coreboot and having decided to pack seabios/grub as a payload will depend on seabios+vbios blob or use seabios graphic init. But that might change. Skulls depends on seabios the same way. linuxboot tries to provide the most generic linux configuration possible (now directly under coreboot since 4.18 if I recall well) and therefore the same problem arises on buildstack dependencies to gnat there, and yet again, if some boards are to depend on libgfxinit+linux to provide VESA, things might fail requiring Nix users to tackle this problem again.

libreml1um, and 4.11 coreboot, could definitely be patched directly under Heads though, if for whatever reason they would want to reenable libgfxinit. Note that libgfxinit offers interesting properties, including drawing a bootsplash directly in ramstage, so that slowing platforms reading from SPI are not left in the dark until linux kernel owns the graphic card for the first time in absence of libgfxinit.

The reason of all of this is purely generic. Ideally, older platforms supported by coreboot 4.11 releaseshould be able to be built if provided with the proper dependencies. And abusing Nix for thatwould simply be amazing. But to be honest, the librem1um being the only known case for libgfxinit potentially being needed might not justify the effort today. The platform is fast enough to not leave users in the dark until i915+drm drivers to own and draw in a functional framebuffer. 

The ecosystem moving forward and away of fb, preferring KMS enabled drm drivers are desired and being pushed forward. So ideally, firwares would present a fb (native init or libgfxinit or efifb) so that the linux kernel can use it without overhead. Ideally, this would mean a single drm driver (simpledrm) with a single fb driver to fallback to (simplefb) which works out of the box without having to pack in the kernel anything big or complicated to get a console output on screen.

When those combined (sandy/ivy bridge platformsused as example here) libgfxinit can draw a bootsplash within millisecond, after which slow spi access continues reading kernel initrd (8 seconds showing bootsplash) and then the kernel kicks in and can use sysfb exposed fb through simplefb. When kexec'ing into another OS, that framebuffer (simplefb) can be exposed through kexec and be reused again through simplefb+simpledrm and then more complex drm drivers can take ownership of the fb again for i915drm. Without libgfxinit, the kernel needs to pack i915drm which even if compressed takes 500kb of precious SPI space. And leave the user in the dark untilthat i915drm driver can draw in the console. This is now 10 seconds on the x230. No problem here, the x230 is currently maintained under coreboot. But the librem1um was supported 2 years ago and now isnt. And the ecosystem is moving leaving that platform not being able to be built with Nix today.

Now, on what really matters here for Nix, ideally, if any of those platforms are deprecated into coreboot tomorrow, the desire is to be able to pin a package commit that would permit that board to continue to be able to build coreboot without any patches needed on top of coreboot. A downstream project using Nix to build a reproducible ROM should be able to do so in 10 years from now. 

Sorry for the long reply and hoping this explains why those old perceived tools are needed otherwise patching things are needed and therefore repruducible builds cannot be obtained (We cannot build librem1um with libgfxinit enabled in coreboot today with Nix).

**Comment by amjoseph-nixpkgs** [comment]
@ofborg eval

**Comment by amjoseph-nixpkgs** [comment]
Rebased.  Apparently github can't see that this is a nonconflicting rebase.

**Comment by yu-re-ka** [comment]
My assumptions:

- Latest coreboot 4.20 will build fine with the more recent gnat we already have in nixpkgs
- coreboot derivatives will eventually move on from coreboot 4.11 to more recent versions

If you want to build coreboot 4.11 from 2019, why don't you use a nixpkgs checkout from 2019, which has the gnat6 you need?

**Comment by amjoseph-nixpkgs** [comment]
> coreboot derivatives will eventually move on from coreboot 4.11 to more recent versions

Bogus assumption.

Coreboot 4.12 dropped support for all the IME/PSP-free hardware.  It was a very contentious decision.  Please don't reopen this old wound.

**Comment by amjoseph-nixpkgs** [comment]
@mmlb, do you have an alternative source for `gnatboot-4.1`?  Gentoo seems to have moved it elsewhere and I can't find the new URL.  I was hoping it would still be in ofborg's cache, but apparently it is not.

```
trying https://ftp.snt.utwente.nl/pub/os/linux/gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
copying path '/nix/store/rhx3c6qvnp8jss777q8znmsifcfy0cw3-source-highlight-3.1.9-dev' from 'https://cache.nixos.org'...
  0   196    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
trying https://distfiles.gentoo.org/distfiles/gnatboot-4.1-amd64.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   302    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
trying https://mirrors.kernel.org/gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
copying path '/nix/store/skylwyykbfnvhq7qssywi0ggwv2ffmm6-python3.10-pyelftools-0.29' from 'https://cache.nixos.org'...
copying path '/nix/store/md052k6fr3ka5p6bbg238qhjrhya1p7j-python3-3.10.12-env' from 'https://cache.nixos.org'...
  0   369    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
error: cannot download source from any mirror
error: builder for '/nix/store/5y0088vk3h5y9hhmcdd55nw52y7irr88-source.drv' failed with exit code 1;
       last 10 log lines:
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0   302    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       >
       > trying https://mirrors.kernel.org/gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0   369    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > error: cannot download source from any mirror
       For full logs, run 'nix log /nix/store/5y0088vk3h5y9hhmcdd55nw52y7irr88-source.drv'.
```

**Comment by mmlb** [comment]
> @mmlb, do you have an alternative source for `gnatboot-4.1`? Gentoo seems to have moved it elsewhere and I can't find the new URL. I was hoping it would still be in ofborg's cache, but apparently it is not.
> 
> ```
> trying https://ftp.snt.utwente.nl/pub/os/linux/gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> copying path '/nix/store/rhx3c6qvnp8jss777q8znmsifcfy0cw3-source-highlight-3.1.9-dev' from 'https://cache.nixos.org'...
>   0   196    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (22) The requested URL returned error: 404
> trying https://distfiles.gentoo.org/distfiles/gnatboot-4.1-amd64.tar.bz2
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0   302    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (22) The requested URL returned error: 404
> trying https://mirrors.kernel.org/gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> copying path '/nix/store/skylwyykbfnvhq7qssywi0ggwv2ffmm6-python3.10-pyelftools-0.29' from 'https://cache.nixos.org'...
> copying path '/nix/store/md052k6fr3ka5p6bbg238qhjrhya1p7j-python3-3.10.12-env' from 'https://cache.nixos.org'...
>   0   369    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (22) The requested URL returned error: 404
> error: cannot download source from any mirror
> error: builder for '/nix/store/5y0088vk3h5y9hhmcdd55nw52y7irr88-source.drv' failed with exit code 1;
>        last 10 log lines:
>        >                                  Dload  Upload   Total   Spent    Left  Speed
>        >   0   302    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>        > curl: (22) The requested URL returned error: 404
>        >
>        > trying https://mirrors.kernel.org/gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2
>        >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>        >                                  Dload  Upload   Total   Spent    Left  Speed
>        >   0   369    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>        > curl: (22) The requested URL returned error: 404
>        > error: cannot download source from any mirror
>        For full logs, run 'nix log /nix/store/5y0088vk3h5y9hhmcdd55nw52y7irr88-source.drv'.
> ```

I do not have one up any where atm but would be fine with a quick GH repo with a release that contains the tarball.

**Comment by mmlb** [comment]
> I feel like I'm still missing answers to some important questions here.
> 
>     1. Would this be a useful step towards obtaining a binary-free bootstrap of modern GNAT? 

Not really, we're still using a binary build of gnat here.

> If so, how far away are we from such a bootstrap? 

Very far it seems. I've recently learned that gnat has *never* been bootstrappable, even the first release required a pre-existing gnat compiler :exploding_head: . There's a project at the _very earliest_ of stages to try and bootstrap gnat (https://fossil.irvise.xyz/gnat-bootstrap/forum)

**Comment by mmlb** [comment]
I've uploaded the gnat4.1 tarball to a repo I control and have updated the PR to use it instead of mirror based off from gentoo. I'd really like to see the repo under nixos' umbrella instead though, thoughts on doing so?

**Comment by RaitoBezarius** [comment]
> I feel like I'm still missing answers to some important questions here.
>  [...]
> 2. > @tlaurion has anybody tried backporting the libgfxinit (the only gnat-using part of coreboot) from a recent coreboot back to 4.11? Since it's written in a completely different language (Ada) than the rest of coreboot (C) the interfaces should be well-defined.

Interestingly, it happens I went into the rabbit hole with Owner Boot.

I am using the coreboot-toolchain 4.16 with coreboot 4.9 (worse than 4.11 arguably) as this is the ownerboot tree (https://git.sr.ht/~amjoseph/ownerboot/).

Basically, for my X230, I encountered many toolchain issues indeed and it was not super hard to fix but quite annoying to use intuition to guess "oh this is a new warning, let's silence it here, let's do that, etc." or read next commits to find out about a new GCC breakage for GNAT (or even GCC!).

So I can confirm it is possible to get:
```
FMAP REGION: FALLBACK
Name                           Offset     Type           Size   Comp
cbfs master header             0x0        cbfs header        32 none
fallback/romstage              0x80       stage           73588 none
fallback/ramstage              0x12080    stage          102052 none
config                         0x2af80    raw               566 none
revision                       0x2b200    raw               570 none
vbt.bin                        0x2b480    raw              1433 LZMA (4281 decompressed)
cmos_layout.bin                0x2ba80    cmos_layout      1804 none
fallback/postcar               0x2c200    stage           15644 none
fallback/dsdt.aml              0x2ff80    raw             13778 none
(empty)                        0x335c0    null          1883672 none
bootblock                      0x1ff400   bootblock        1968 none
FMAP REGION: NORMAL
Name                           Offset     Type           Size   Comp
cbfs master header             0x0        cbfs header        32 none
fallback/romstage              0x80       stage           73588 none
fallback/ramstage              0x12080    stage          102052 none
config                         0x2af80    raw               566 none
revision                       0x2b200    raw               570 none
vbt.bin                        0x2b480    raw              1433 LZMA (4281 decompressed)
cmos_layout.bin                0x2ba80    cmos_layout      1804 none
fallback/postcar               0x2c200    stage           15644 none
fallback/dsdt.aml              0x2ff80    raw             13778 none
(empty)                        0x335c0    null          1885716 none
installing
```

Of course, it fails for unrelated reason I believe:
```
E: Image is missing 'COREBOOT' region
E: Could not add [/nix/store/a22y44d8p7pzv93kjgl5by0hv6fwn2a4-linux-x86_64-unknown-linux-gnu-5.10.148/bzImage, 7506985 bytes (7331 KB)@0x0]; too big?
E: Failed to add '/nix/store/a22y44d8p7pzv93kjgl5by0hv6fwn2a4-linux-x86_64-unknown-linux-gnu-5.10.148/bzImage' into ROM image.
```

Maybe, I will get somewhere on a future nerdsnipe in some oceansprint.org with some fellows who are interested into better coreboot in nixpkgs.

As a matter of fact, I believe it is very possible to use libgfxinit + libhwbase on older coreboot with newer coreboot-toolchain (4.16), I think we can go further, but I will let this to the next one, the only reason I went for 4.16 is that glibc changed the fact that minimum stack size was a constant, and it created serious issues with GCC 8.3.0.

**Comment by mmlb** [comment]
no longer needing this, so just going to go ahead on close

**Pull request closed without merging**